src/libcore/hash/mod.rs

```diff
@@ -691,6 +692,8 @@ mod impls {
                 state.write_usize(*self as *const () as usize);
             } else {
                 // Fat pointer
+                // SAFETY: we are accessing the memory occupied by `self`
+                // which is guaranteed to be valid.
                 let (a, b) = unsafe { *(self as *const Self as *const (usize, usize)) };
```

Comment:
Ah... well, this seems like a bit of an assumption actually, that only std is permitted to make (I believe) -- we don't guarantee that DSTs are (usize, usize) tuples so we should note that here.

(This is safe to do in std, but to my knowledge not elsewhere).

Reply:
I don't really understand why it would be safe only in std. `T` may be any type, inside or outside of the standard library? I don't think it would help much to just say "this is only safe in std" in the safety comment, right?

Reply:
It's an implementation detail of the compiler. Since `std` is shipped with the compiler, it can always be kept in sync with such details. Outsiders can't make assumptions like this.

Reply:
Ohhh, I see what you mean. Today DSTs are `(usize, usize)` but it isn't guaranteed that it will be the case in the future. Thanks for the clarifications!